keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+### Goals
+
+* Provide structured logging interface in K8s codebase
+* Standardize commonly used logging keys (e.g. referencing k8s object, http request fields)
+* Propose sound strategy for migration ensuring its completeness
+* Make implementation of context based solutions like logs correlation or tracing within Kubernetes easier.
+
+### Non-Goals
+
+* We are **not** changing logging pipeline in K8s
+* We are **not** standardizing some specific log messages
+* We are **not** tackling logging keys lifecycle (should be addressed by separate KEP when new API is close to GA)
+
+## Proposal
+
+This KEP proposes introducing a structured logging into Kubernetes control-plane components. As part of this proposal we will introduce a new logging API that will replace klog. We will introduce a Kubernetes logging schema which will provide a way to define a common set of logging keys and ensure their consistent usage. To provide immediate benefit for adapting a new API we will propose a new logging format that will simplify ingestion of structured data into third party logging solutions. As part of changes in kubernetes logging we propose to tightly couple usage of logging and golang context.
+
+### Logging Interface
+
+When choosing new structured logging API following criteria has been used:
+* Selected API should provide abstract structured logging interface so that it would be easy to inject different implementations for the same API.
+* Selected API should be minimalistic so that it should be easy to reimplement, wrap or to mock it.
+* Applicability of the selected API should not be limited to the Kubernetes control plane but it should be possible to use the same API in other components running on Kubernetes clusters which are also implemented in go.
+* Selected API should support structured logging by providing a way to define a set of key-value pairs for each log entry.
+* There should at least one production ready implementation of the selected API which could produce JSON formatted logs.
+* It should be possible to produce klog compatible logs from the selected API which could be used during the migration to a new API.
+
+Based on this criteria https://github.com/go-logr/logr log interface has been selected as a logging API for Kubernetes.
+
+Apart from meeting the criteria listed above logr has a few nice features:
+* Constant log message together with logger name provide stable identification for log entries which greatly simplifies querying logs.
+* Limiting log levels to just Info and Error makes it much easier to choose the proper log level for given log.
+* Using key-values pairs for defining log fields allows more efficient implementation of the logger.
+
+#### Logger initialization
+
+Logger implementation initialization should be done at startup of component and passed down the stack as an interface. Aside of initialization of logger, there should not be any dependency on specific implementation of logger Interface.
```

Comment:
How much can be changed at runtime?  We change glog levels at runtime today.  Many times, a server can get into a weird state and being able to change the configuration at runtime has value.

Reply:
This will depend on the actual implementation used by logr but apart from some performance overhead related to synchronization I do not see a major problem of retaining this dynamic log level functionality with new log output format.